src/libpriv/rpmostree-core.c

```diff
@@ -2854,8 +2870,37 @@ delete_package_from_root (RpmOstreeContext *self,
       if (!g_str_has_prefix (fn, "usr/"))
         continue;
 
-      if (!glnx_shutil_rm_rf_at (rootfs_dfd, fn, cancellable, error))
-        return FALSE;
+      /* Delete files first, we'll handle directories next */
+      if (S_ISDIR (mode))
+        g_sequence_insert_sorted (dirs_to_remove, g_strdup (fn), compare_strlen, NULL);
+      else
+        {
+          if (unlinkat (rootfs_dfd, fn, 0) < 0)
+            {
+              if (errno != ENOENT)
+                return glnx_throw_errno_prefix (error, "unlinkat(%s)", fn);
+            }
+        }
+    }
+
+  /* We want to perform a bottom-up removal of directories. This is implemented
+   * by having a sequence of filenames sorted by decreasing length.  A filename
+   * of greater length must be a subdirectory, or unrelated.
+   *
+   * If a directory isn't empty, that's fine - a package may own
+   * the directory, but it's valid for other packages to put content under it.
+   * Particularly while we're doing a replacement.
+   */
+  GSequenceIter *iter = g_sequence_get_begin_iter (dirs_to_remove);
+  while (!g_sequence_iter_is_end (iter))
+    {
+      const char *fn = g_sequence_get (iter);
+      if (unlinkat (rootfs_dfd, fn, AT_REMOVEDIR) < 0)
+        {
+          if (!G_IN_SET (errno, ENOENT, EEXIST, ENOTEMPTY))
```

Comment:
Should we add `ENOTDIR` here? In case a parent already doesn't exist somehow?

Reply:
`glnx_shutil_rm_rf_at()` currently throws an error in that case.  Did you figure out what librpm does?

I'm not opposed to ignoring it...but every case I can think of right now it would happen feels more like an error.

Reply:
>  Did you figure out what librpm does?

Yeah, it ignores `ENOENT` and `ENOTEMPTY` only. Matching that behaviour as well sounds good to me!